ParProcCo/job_scheduler.py

```diff
@@ -0,0 +1,274 @@
+from drmaa2 import JobSession, JobTemplate, Drmaa2Exception
+
+
+class TemporarilyLogToFile:
+
+    def __init__(self, logger, path=None):
+        self.logger = logger
+        self.log_path = path
+        if self.log_path is not None:
+            self.file_handler = logging.FileHandler(self.log_path, "a+")
+            self.logger.addHandler(self.file_handler)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        self.close()
+
+    def close(self):
+        if self.log_path is not None:
+            self.file_handler.close()
+            self.logger.removeHandler(self.file_handler)
+            with open(self.log_path, "a+") as fp:
+                fp.write("\n")
+
+
+class JobScheduler:
+
+    def __init__(self, working_directory, cluster_output_dir, project, priority, cpus=16, timeout=timedelta(hours=2)):
+        """
+        JobScheduler can be used for cluster job submissions
+        Args:
+            working_directory (pathlib.Path or str): working directory
+            cluster_output_dir (pathlib.Path or str): cluster output directory
+            project (str): project name
+            priority (str): name of queue to submit to
+            cpus (int, optional): Number of CPUs to request. Defaults to 16.
+            timeout (int, optional): Timeout for cluster jobs in minutes. Defaults to 180 (2 hours).
+        """
+        self.working_directory = Path(working_directory)
+        self.cluster_output_dir = Path(cluster_output_dir)
+        self.project = self.check_project_list(project)
+        self.priority = self.check_queue_list(priority)
+        self.cpus = cpus
+        self.timeout = timeout
+        self.logger = logging.getLogger()
+        self.batch_number = 0
+        self.output_paths = []
+        self.start_time = datetime.now()
+        self.job_history = {}
+        self.job_completion_status = {}
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc):
+        pass
+
+    def check_queue_list(self, priority):
+        if priority == "" or priority is None:
+            raise ValueError(f"priority is empty")
+        priority = priority.lower()
+        q_proc = os.popen('qconf -sql')
+        q_name_list = q_proc.read().split()
+        q_proc.close()
+        if priority in q_name_list:
+            return priority
+        else:
+            raise ValueError(f"priority {priority} not in queue list: {q_name_list}\n")
+
+    def check_project_list(self, project):
+        if project == "" or project is None:
+            raise ValueError(f"project is empty")
+        prj_proc = os.popen('qconf -sprjl')
+        prj_name_list = prj_proc.read().split()
+        prj_proc.close()
+        if project in prj_name_list:
+            return project
+        else:
+            raise ValueError(f"{project} not in list of project names: {prj_name_list}\n")
+
+    def check_jobscript(self, jobscript):
+        if not Path(jobscript).exists():
```

Comment:
Can make it a path object and return that. Want to check it's a file and then try opening it.